src/systems/systems/src/shared/can_bus.rs

```diff
@@ -0,0 +1,516 @@
+                        self.next_output_message_valid = true;
+
+                        break;
+                    }
+                }
+
+                self.next_transmitting_system = (self.next_transmitting_system + 1) % N;
+            }
+        } else {
+            self.message_received_by_systems
+                .iter_mut()
+                .for_each(|received| *received = [true, true]);
+        }
+    }
+
+    pub fn new_message_received(&self, function_id: u8) -> bool {
+        for (i, id) in self.attached_systems.iter().enumerate() {
+            if *id == function_id {
+                return !self.message_received_by_systems[i][0];
+            }
+        }
+
+        false
+    }
+
+    pub fn received_message(&mut self, function_id: u8) -> Arinc825Word<f64> {
+        let mut system_idx = N;
+        for (i, id) in self.attached_systems.iter().enumerate() {
+            if *id == function_id {
+                system_idx = i;
+                break;
+            }
+        }
+
+        if system_idx < N {
+            self.message_received_by_systems[system_idx] = [true, true];
+        }
+
+        self.received_message
+    }
+
+    pub fn send_message(&mut self, message: Arinc825Word<f64>) -> bool {
+        for (i, id) in self.attached_systems.iter().enumerate() {
+            if *id == message.client_function_id() {
+                // detected an buffer overrun
+                if self.transmission_buffers[i].len() >= TRANSMISSION_BUFFER_SIZE {
+                    return false;
+                }
+
+                self.transmission_buffers[i].push_back(message);
+                return true;
+            }
+        }
+
+        false
+    }
+
+    pub fn reset_buffer(&mut self, function_id: u8) {
+        for (i, id) in self.attached_systems.iter().enumerate() {
+            if *id == function_id {
+                self.transmission_buffers[i] = VecDeque::new();
+                break;
+            }
+        }
+    }
+}
+
+impl<const N: usize> SimulationElement for CanBus<N> {
+    fn read(&mut self, reader: &mut SimulatorReader) {
+        self.received_message = reader.read_arinc825(&self.databus_id);
+
+        let availability: f64 = reader.read(&self.availability_id);
+        self.available = availability != 0.0;
+
+        for (i, id) in self.message_received_by_systems_ids.iter().enumerate() {
+            if !self.message_received_by_systems[i][1] {
+                let message_received: f64 = reader.read(id);
+                self.message_received_by_systems[i][0] = message_received != 0.0;
+            }
+        }
+
+        let failure: f64 = reader.read(&self.failure_indication_id);
+        self.failure_indication = failure != 0.0;
+    }
+
+    fn write(&self, writer: &mut SimulatorWriter) {
+        // set the availability flag
+        writer.write(
+            &self.availability_id,
+            if self.failure_indication { 0. } else { 1. },
```

Comment:
```suggestion
            if self.failure_indication { 0. } else { 1. },
```
I've realised that ``writer.write(&self.availability_id,, !self.failure_indication)`` would probably work too, if you want to test/implement that you can, but otherwise I think this is readable enough.